[2.13.x] DDF-4051 Permissions granted to ApplicationService to install new features should be granted to other bundles

lessarderic opened the pull request:
#### What does this PR do?
Adds AccessController.doPrivilegedAction() call to ApplicationServiceBean's installFeature() method to grant same feature install permissions to other bundles.

#### Who is reviewing it? 
<!--(please choose AT LEAST two reviewers that need to approve the PR before it can get merged)-->
@brjeter 
@clockard

#### Select relevant component teams: 
https://github.com/orgs/codice/teams/security
#### Ask 2 committers to review/merge the PR and tag them here.
@coyotesqrl
@stustison

#### How should this be tested?
Unzip the distribution (or any downstream distribution), start karaf and make sure that the installer UI works as expected, especially when the profile selected contains a feature that needs to install a file or files under `/etc`.

Since this is an intermittent problem, the test should be repeated multiple times on both Windows and Linux.

#### Any background context you want to provide?

#### What are the relevant tickets?
[DDF-4051](https://codice.atlassian.net/browse/DDF-4051)

#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

coyotesqrl commented on platform/admin/core/admin-core-appservice/src/main/java/org/codice/ddf/admin/application/service/impl/ApplicationServiceBean.java:
```diff
@@ -181,8 +183,13 @@ public void destroy() throws ApplicationServiceException {
   @Override
   public void installFeature(String feature) {
```
✏️ I think we should privilege the `uninstallFeature` method as well. I'm going to link to a gist with my suggested version of this class on this review.

lessarderic commented:
build now

lessarderic commented:
build now

lessarderic commented:
build now

coyotesqrl commented on platform/security/platform-security-core-api/src/main/java/ddf/security/samlp/SimpleSign.java:
```diff
@@ -13,8 +13,8 @@
 
-import static com.sun.org.apache.xml.internal.security.signature.XMLSignature.ALGO_ID_SIGNATURE_DSA_SHA256;
 import static org.apache.commons.lang.CharEncoding.UTF_8;
+import static org.opensaml.xmlsec.signature.support.SignatureConstants.ALGO_ID_SIGNATURE_DSA_SHA256;
```
👍

brjeter commented:
✅ Herod successfully, installing 15 times on Windows 2012 and 10 times on CentOS 7.

lessarderic commented:
build now

brjeter commented:
build now (failed on UI building)

paouelle commented on platform/admin/core/admin-core-appservice/src/test/java/org/codice/ddf/admin/application/service/impl/ApplicationServiceBeanTest.java:
```diff
@@ -236,6 +244,48 @@ public void testInstallProfile() throws Exception {
+
+    assertThat(
+        Arrays.stream(stackTraceElements.get())
+            .anyMatch(e -> e.toString().startsWith(DO_PRIVILEGED_STACK_ELEMENT)),
```
❗️ This relies on the `toString()` returning a given format. You would be better of testing the classname separately. As for testing the test method, I think it is sufficient to check if the AccessController is on the stack.

paouelle commented on platform/admin/core/admin-core-appservice/src/test/java/org/codice/ddf/admin/application/service/impl/ApplicationServiceBeanTest.java:
```diff
@@ -236,6 +244,48 @@ public void testInstallProfile() throws Exception {
+    serviceBean.installFeature("profile-name");
+
+    assertThat(
+        Arrays.stream(stackTraceElements.get())
```
❗️ The problem here is that you have no way to know if the code that starts unit testing and ends up calling your test method doesn't already have the AccessController in the stack. The best approach would be to verify up the stack until you hit your test method and doing this inside a test case is not necessarily a good idea. I think this would be better done in a JUnit rule which would have access to the test method. You could then record the stack at the beginning and compare it with the stack at the end to make sure the AccessController is in the stack. While only doing this if the test method is annotated to check access controller. That would be reusable.

lessarderic commented:
@paouelle, I updated the PR based on your comments and extracted some utility classes to make things look a little better.

If we feel good about the approach, I'll move those inner classes to the `test-common` library so I can reuse them elsewhere.

clockard commented:
build now

paouelle commented on platform/admin/core/admin-core-appservice/src/test/java/org/codice/ddf/admin/application/service/impl/ApplicationServiceBeanTest.java:
```diff
@@ -562,4 +608,57 @@ public void testGetPluginsForApplication() throws Exception {
+  public static class StackCaptor {
+    AtomicReference<StackTraceElement[]> stackTraceElements = new AtomicReference<>();
+
+    public Stubber doCaptureStack() {
```
✏️ Here is an idea ... capture the starting stack when doCapture() is called such that you can remove the first matching frames from your search in your matcher later.
